saleor/checkout/complete_checkout.py

```diff
@@ -0,0 +1,178 @@
+from .utils import abort_order_data, create_order, get_order, prepare_order_data
+
+
+@transaction.atomic
+def prepare_checkout(
+    checkout: models.Checkout, discounts, tracking_code, redirect_url, payment
+):
+    lines = list(checkout)
+
+    clean_checkout_shipping(checkout, lines, discounts, CheckoutErrorCode)
+    clean_checkout_payment(
+        checkout, lines, discounts, CheckoutErrorCode, last_payment=payment
+    )
+
+    if redirect_url:
+        try:
+            validate_storefront_url(redirect_url)
+        except ValidationError as error:
+            raise ValidationError(
+                {"redirect_url": error}, code=AccountErrorCode.INVALID.value
+            )
+
+    to_update = []
+    if redirect_url and redirect_url != checkout.redirect_url:
+        checkout.redirect_url = redirect_url
+        to_update.append("redirect_url")
+
+    if tracking_code and tracking_code != checkout.tracking_code:
+        checkout.tracking_code = tracking_code
+        to_update.append("tracking_code")
+
+    if to_update:
+        to_update.append("last_change")
+        checkout.save(update_fields=to_update)
+
+
+@transaction.atomic
+def convert_checkout_to_order(
+    checkout: models.Checkout, order_data: dict, user: Optional[User]
+) -> order_models.Order:
+    order = create_order(
+        checkout=checkout, order_data=order_data, user=user,  # type: ignore
+    )
+    # remove checkout after order is successfully created
+    checkout.delete()
+    return order
+
+
+def get_order_data(checkout: models.Checkout, discounts: List[DiscountInfo]) -> dict:
+    try:
+        with transaction.atomic():
+            order_data = prepare_order_data(
+                checkout=checkout, lines=list(checkout), discounts=discounts,
+            )
+    except InsufficientStock as e:
+        raise ValidationError(f"Insufficient product stock: {e.item}", code=e.code)
+    except NotApplicable:
+        raise ValidationError(
+            "Voucher not applicable",
+            code=CheckoutErrorCode.VOUCHER_NOT_APPLICABLE.value,
+        )
+    except TaxError as tax_error:
+        raise ValidationError(
+            "Unable to calculate taxes - %s" % str(tax_error),
+            code=CheckoutErrorCode.TAX_ERROR.value,
+        )
+    return order_data
+
+
+def process_payment(
+    payment: Payment,
+    store_source: bool,
+    payment_data: Optional[dict],
+    order_data: dict,
+) -> Transaction:
+    try:
+        if payment.to_confirm:
+            txn = gateway.confirm(payment, additional_data=payment_data)
+        else:
+            txn = gateway.process_payment(
+                payment=payment,
+                token=payment.token,
+                store_source=store_source,
+                additional_data=payment_data,
+            )
+        payment.refresh_from_db()
+        if not txn.is_success:
+            raise PaymentError(txn.error)
+    except PaymentError as e:
+        abort_order_data(order_data)
```

Comment:
`abort_order_data` is also only used for completing checkouts, can we move it to this module? Also, this name is really awkward, we could fix it on this occasion.

Reply:
Changed